File: implementation/backend/src/GoalExtraction.Domain/Entities/Goal.cs
namespace GoalExtraction.Domain.Entities;

using System.ComponentModel.DataAnnotations;
using GoalExtraction.Domain.Common;
using GoalExtraction.Domain.Enums;

public class Goal : BaseEntity
{
    [Required]
    public Guid EmployeeId { get; set; }

    [System.Text.Json.Serialization.JsonIgnore]
    public Employee? Employee { get; set; }

    [Required]
    [MaxLength(120)]
    public string Title { get; set; } = string.Empty;

    [Required]
    [MaxLength(1000)]
    public string Description { get; set; } = string.Empty;

    public GoalCategory Category { get; set; } = GoalCategory.PERFORMANCE;

    [MaxLength(500)]
    public string Metric { get; set; } = string.Empty;

    [MaxLength(100)]
    public string Timeframe { get; set; } = string.Empty;

    public GoalPriority Priority { get; set; } = GoalPriority.MEDIUM;

    public GoalStatus Status { get; set; } = GoalStatus.ACTIVE;

    public GoalProvenance Provenance { get; set; } = GoalProvenance.AI_ORIGINAL;

    [MaxLength(64)]
    public string SourceTranscriptHash { get; set; } = string.Empty;

    public Guid? ReviewerId { get; set; }
}

File: implementation/frontend/src/types/goal.ts
export type GoalCategory = 'PERFORMANCE' | 'DEVELOPMENT' | 'PROJECT' | 'TECHNICAL';

export type GoalPriority = 'HIGH' | 'MEDIUM' | 'LOW';

export type GoalProvenance = 'AI_ORIGINAL' | 'AI_MODIFIED' | 'MANUAL';

export interface ProposedGoal {
  tempId: string;
  title: string;
  description: string;
  category: GoalCategory;
  metric: string;
  timeframe: string;
  priority: GoalPriority;
  provenance: GoalProvenance;
}

export interface EditableGoal extends ProposedGoal {
  localId: string;
  isEdited: boolean;
  validationErrors: Record<string, string>;
}

export interface Employee {
  id: string;
  fullName: string;
  email: string;
  department: string;
}

export const GOAL_CATEGORIES: GoalCategory[] = [
  'PERFORMANCE',
  'DEVELOPMENT',
  'PROJECT',
  'TECHNICAL',
];

export const GOAL_PRIORITIES: GoalPriority[] = ['HIGH', 'MEDIUM', 'LOW'];

export function validateGoal(goal: Partial<ProposedGoal>): Record<string, string> {
  const errors: Record<string, string> = {};

  const title = (goal.title || '').trim();
  if (!title) {
    errors.title = 'Title cannot be empty.';
  } else if (title.length < 5 || title.length > 120) {
    errors.title = 'Title must be between 5 and 120 characters.';
  }

  const desc = (goal.description || '').trim();
  if (!desc) {
    errors.description = 'Description cannot be empty.';
  } else if (desc.length < 10 || desc.length > 1000) {
    errors.description = 'Description must be between 10 and 1000 characters.';
  }

  const metric = (goal.metric || '').trim();
  if (metric.length > 500) {
    errors.metric = 'Metric cannot exceed 500 characters.';
  }

  const timeframe = (goal.timeframe || '').trim();
  if (timeframe.length > 100) {
    errors.timeframe = 'Timeframe cannot exceed 100 characters.';
  }

  return errors;
}
